### ISSUE TITLE
Request: CodeGen output in relative directories

### ISSUE THREAD

**Issue by AnthonyMDev** [apollographql/apollo-ios#1083]
It looks like for TypeScript/Flow, the code gen tooling supports relative directories. (Generated files are located in a directory relative to the directory of the source .graphql file) Is there any chance we can get this implemented for Swift?

**Comment by designatednerd** [comment]
Definitely a chance. I'll stipulate that it probably won't be as part of the initial version but given that we get the file path of the file being generated as part of the IR, this shouldn't be too awful.*

* - famous last words

**Comment by designatednerd** [comment]
Spoke with @AnthonyMDev on this and he's willing to take a stab. 

Notes: 
- [`ASTFragment`](https://github.com/apollographql/apollo-ios/blob/master/Sources/ApolloCodegenLib/ASTFragment.swift) and [`ASTOperation`](https://github.com/apollographql/apollo-ios/blob/master/Sources/ApolloCodegenLib/ASTOperation.swift) both have a `filePath` property which should give you the string path where something is located. 
- There isn't one for [`ASTTypeUsed`](https://github.com/apollographql/apollo-ios/blob/master/Sources/ApolloCodegenLib/ASTTypeUsed.swift), which covers enums and Input Objects, so we'd have to figure out where to put those. 
- You'll probably want to make this an option on the [`ApolloCodegenOptions.OutputFormat`](https://github.com/apollographql/apollo-ios/blob/master/Sources/ApolloCodegenLib/ApolloCodegenOptions.swift#L7) enum.

**Comment by AnthonyMDev** [comment]
Oh good point. The `Types.graphql.swift` file needs somewhere to live... Ugh. Any ideas on this?

I'm wondering if Types should just be declared in the module they are used in? If multiple modules are using a type, then perhaps there needs to be some lowest level module everything imports? That's kinda gross too.

**Comment by designatednerd** [comment]
We could use namespaces, but that'd be a hot mess express of duplicated code.....

🤔 let me think on it (oh shit i touched my face)

**Comment by AnthonyMDev** [comment]
Yeah, I'll continue thinking about this as well. I've got another project I need to finish up, and then this is the next thing I'll be working on after that.

**Comment by designatednerd** [comment]
Awesome. HUGE thanks for your help.

**Comment by AnthonyMDev** [comment]
I'm not sure if there is a great universal way to solve the "TypesUsed" issue. I'm wondering if, as a simple measure would be to just allow the user an option to specifically set the path for that generated file and a name of the module that includes it? That way people could do what they want with it. 

For example, with me, using CocoaPods, I have a "Core" pod that includes stuff that is common to basically all my modules. (Stuff like the Logger/CrashReporting, Dependency Resolution, and some app wide constants.) I could just dump it in there.

When you call code-gen, you have to also provide the name of the module for common files like Types.graphql.swift (and if you have Custom Scalar typealiases you can put those in there also). Then ALL generated GraphQL files can just import that module and call it a day?

Just like Cocoapods, this could work with Carthage, SPM, or other setups and you would just have to do some manual setup of your project structure, but that's nothing new.

What do you think @designatednerd ?

**Comment by designatednerd** [comment]
Yeah, I think the "relative directories" option can have plenty of associated values - this is definitely one that's for more advanced users.

**Comment by gsabran** [comment]
FWIW here's how I've approached this (for Airbnb):

- For fragments/operations, the codegen uses a config that allows to define a hook to resolve the codegen destination from the GQL definition location. I'm simply using it as `<path>.graphql` -> `<path>.codegen.swift`, but this is very configurable.
- For enums/input types we:
  - keep them in any of the codegened files that uses them if they are only referenced by a single module. This works for the majority of cases.
  - else the codegen config specifies a list of rules that can be used to deambiguate, ending with a default rule. For instance:
```js
[
  { type: 'User.*', destination: './DataModels/User/{{typeName}}.codegen.swift' },
  { type: '.*', destination: './SharedModule/Sources/{{typeName}}.codegen.swift' },
]
```
- Finally we map the GQL dependencies to those now defined file dependencies, and understand what import statements need to be generated.

**Comment by AnthonyMDev** [comment]
Put up a proposal for how to do this. #1223

**Comment by designatednerd** [comment]
@AnthonyMDev I think your proposal handles output for a multi-module schema, I think there's a good case to also allow output in relative directories for a single module. 

In a reasonably large single-module codebase, people can to keep the `.swift` files in the same folder as the `.graphql` files that declare them. Especially for people who are keeping the `.graphql` files near the point of use, this could be super-helpful. 

We'd probably still need to ask users to declare a default output path, but I think that's actually not going to be too bad with the Swift Codegen (like all things it would be a goddamned nightmare with the TS codegen).

**Comment by calvincestari** [comment]
Hi 👋🏻 - this issue should be resolved in 1.0. 1.0 is currently in alpha releases with a full 1.0 release in H2 2022. Please take a look at the [latest alpha](https://github.com/apollographql/apollo-ios/releases) and let us know if this is still an issue for you.
